### PR TITLE
Add weak referenced type `WeakAddr<A>` for actors

### PR DESCRIPTION
**Pull request by feisuzhu**
Fixes https://github.com/actix/actix/issues/187

### CODE REVIEW

File: src/address/channel.rs [hunk 1]
```diff
@@ -498,6 +518,21 @@ impl<A: Actor> Hash for AddressSender<A> {
...
 }
 
+//
+//
+// ===== impl WeakSender =====
+//
+//
+impl<A: Actor> WeakAddressSender<A> {
+    /// TODO
+    pub fn upgrade(&self) -> Option<AddressSender<A>> {
```
[the remark this answers is not in the document]
Comment (feisuzhu) [reply src/address/channel.rs]: @axos88 `std::rc::Weak` does the same. And there is no other reason for a failed upgrade :)

File: src/address/mod.rs [hunk 2]
```diff
@@ -165,6 +172,27 @@ impl<A: Actor> Hash for Addr<A> {
...
+/// Weak referenced Addr<A>
+#[derive(Debug)]
+pub struct WeakAddr<A: Actor> {
+    wtx: WeakAddressSender<A>,
+}
+
+impl<A: Actor> WeakAddr<A> {
+    pub fn upgrade(&self) -> Option<Addr<A>> {
+        match self.wtx.upgrade() {
+            Some(tx) => {
```
[the remark this answers is not in the document]
Comment (feisuzhu) [reply src/address/mod.rs]: I can't……
It says I'm moving `tx` into guarding expr thus can't use it later.

[the remark this answers is not in the document]
Comment (DoumanAsh) [reply src/address/mod.rs]: Ah ok.... sorry, ignore it

File: tests/test_address.rs [hunk 3]
```diff
@@ -69,12 +76,46 @@ fn test_address() {
...
+
+    fn started(&mut self, ctx: &mut Self::Context) {
+        let addr1 = MyActor3.start();
+        let addr2 = MyActor3.start();
+        let weak1 = addr1.downgrade();
+        let weak2 = addr2.downgrade();
+        drop(addr1);
+
+        ctx.run_later(Duration::new(0, 1000), move |_, _| {
+            match weak1.upgrade() {
```
[the remark this answers is not in the document]
Comment (feisuzhu) [reply tests/test_address.rs]: fixed :)

File: src/address/channel.rs [hunk 4]
```diff
@@ -498,6 +518,20 @@ impl<A: Actor> Hash for AddressSender<A> {
     }
 }
 
+//
+//
+// ===== impl WeakSender =====
+//
+//
+impl<A: Actor> WeakAddressSender<A> {
+    pub fn upgrade(&self) -> Option<AddressSender<A>> {
```
Comment (DoumanAsh) [inline src/address/channel.rs]: Add doc comment

File: src/address/mod.rs [hunk 5]
```diff
@@ -165,6 +172,27 @@ impl<A: Actor> Hash for Addr<A> {
...
 }
 
+/// Weak referenced Addr<A>
+#[derive(Debug)]
+pub struct WeakAddr<A: Actor> {
+    wtx: WeakAddressSender<A>,
+}
+
+impl<A: Actor> WeakAddr<A> {
+    pub fn upgrade(&self) -> Option<Addr<A>> {
```
Comment (DoumanAsh) [inline src/address/mod.rs]: Add doc comment

**Pull request merged**